### PR TITLE
kubeadm: validate podSubnet against node-mask and serviceSubnet max size

### PR DESCRIPTION
**Pull request by aojea**
the controller manager should validate the podSubnet against the node-mask
because if they are incorrect can cause the controller-manager to fail.

We don't need to calculate the node-cidr-masks, because those should
be provided by the user, if they are wrong we fail in validation.

The Service subnet is also limited in size for technical reasons, 
so we validate that the subnet size does not exceed the limits.

xref: https://github.com/kubernetes/enhancements/pull/1881

/kind cleanup

```release-note
kubeadm: improve the validation of serviceSubnet and podSubnet.
ServiceSubnet has to be limited in size, due to implementation details, and the mask can not allocate more than 20 bits.
PodSubnet validates against the corresponding cluster "--node-cidr-mask-size" of the kube-controller-manager, it fail if the values are not compatible.
ACTION REQUIRED: kubeadm no longer sets the node-mask automatically on IPv6 deployments, you must check that your IPv6 service subnet mask is compatible with the default node mask /64 or set it accordenly. 
Previously, for IPv6, if the podSubnet had a mask lower than /112, kubeadm calculated a node-mask to be multiple of eight and splitting the available bits to maximise the number used for nodes.
```

### CODE REVIEW

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go [hunk 1]
```diff
@@ -403,21 +403,45 @@ func ValidateIPNetFromString(subnetStr string, minAddrs int64, isDualStack bool,
...
+		} else if (nodeMask - maskSize) > constants.PodSubnetNodeMaskMaxDiff {
+			allErrs = append(allErrs, field.Invalid(fldPath, podSubnet.String(), "pod subnet mask and node-mask different is greater than 16"))
+		}
+	}
+	return allErrs
+}
+
 // ValidateNetworking validates networking configuration
-func ValidateNetworking(c *kubeadm.ClusterConfiguration, fldPath *field.Path) field.ErrorList {
+func ValidateNetworking(cfg *kubeadm.ClusterConfiguration, fldPath *field.Path) field.ErrorList {
```
Comment (aojea) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go:429]: is an unrelated change but using the  `c`  only was problematic to me during the refactor, I think that this  reads better

File: cmd/kubeadm/app/phases/controlplane/manifests.go [hunk 2]
```diff
@@ -361,29 +315,6 @@ func getControllerManagerCommand(cfg *kubeadmapi.ClusterConfiguration) []string
 		defaultArguments["cluster-name"] = cfg.ClusterName
 	}
 
-	// TODO: The following code should be remvoved after dual-stack is GA.
-	// Note: The user still retains the ability to explicitly set feature-gates and that value will overwrite this base value.
-	enabled, present := cfg.FeatureGates[features.IPv6DualStack]
-	if present {
-		defaultArguments["feature-gates"] = fmt.Sprintf("%s=%t", features.IPv6DualStack, enabled)
```
Comment (aojea) [inline cmd/kubeadm/app/phases/controlplane/manifests.go:368]: does this makes sense?
we already have the config with the feature gate enabled

File: cmd/kubeadm/app/phases/controlplane/manifests.go [hunk 3]
```diff
@@ -361,29 +315,6 @@ func getControllerManagerCommand(cfg *kubeadmapi.ClusterConfiguration) []string
...
-	}
-	if cfg.Networking.PodSubnet != "" {
```
Comment (aojea) [inline cmd/kubeadm/app/phases/controlplane/manifests.go:370]: since we validate that the node-masks are correct , we must not change the masks later

File: cmd/kubeadm/app/componentconfigs/controllermanager.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+	defaultNodeMaskCIDRIPv4 = 24
+
+	// DefaultNodeMaskCIDRIPv6 is default mask size for IPv6 node cidr for use by the controller manager
+	defaultNodeMaskCIDRIPv6 = 64
+)
+
+// GetClusterNodeMask return the corresponding node-cidr-mask
+// based on the Cluster configuration and the IP family
+// Default is 24 for IPv4 and 64 for IPv6
+func GetClusterNodeMask(cfg *kubeadmapi.ClusterConfiguration, isIPv6, isDualStack bool) int {
```
Comment (neolit123) [inline cmd/kubeadm/app/componentconfigs/controllermanager.go:36]: can we add GetClusterNodeMask in validation.go and lowercase it given it is used only there?

ideally we should not have this file cmd/kubeadm/app/componentconfigs/controllermanager.go, because the componentconfigs package is reserved for kubelet and kube-proxy for the time being.

File: cmd/kubeadm/app/componentconfigs/controllermanager.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
+)
+
+const (
+	// defaultNodeMaskCIDRIPv4 is default mask size for IPv4 node cidr for use by the controller manager
+	defaultNodeMaskCIDRIPv4 = 24
+
+	// DefaultNodeMaskCIDRIPv6 is default mask size for IPv6 node cidr for use by the controller manager
+	defaultNodeMaskCIDRIPv6 = 64
```
Comment (neolit123) [inline cmd/kubeadm/app/componentconfigs/controllermanager.go:30]: this const block can be inside the function bellow.

File: cmd/kubeadm/app/constants/constants.go [hunk 6]
```diff
@@ -199,6 +199,15 @@ const (
 	// We need at least ten, because the DNS service is always at the tenth cluster clusterIP
 	MinimumAddressesInServiceSubnet = 10
 
+	// MinimumAddressesInPodSubnet defines minimum amount of pods the Pod subnet should allow.
```
Comment (neolit123) [inline cmd/kubeadm/app/constants/constants.go:202]: `of pods` -> `of Pods`

File: cmd/kubeadm/app/constants/constants.go [hunk 7]
```diff
@@ -199,6 +199,15 @@ const (
...
+	// We need at least more than services, an IPv4 /28 or IPv6 /128 subnet means 14 util addressess
+	MinimumAddressesInPodSubnet = 14
+
+	// The nodes subnet mask size cannot be greater than 16 more than the pod subnet mask size
+	// xref: https://github.com/kubernetes/kubernetes/issues/44918
+	// PodSubnetNodeMaskMaxDiff limited to 16 due to the uncompressed bitmap
+	PodSubnetNodeMaskMaxDiff = 16
```
Comment (neolit123) [inline cmd/kubeadm/app/constants/constants.go:209]: comment above should start with PodSubnetNodeMaxDiff on the first line.

i tried to reword this, please change if needed.

> // PodSubnetNodeMaskMaxDiff is limited to 16 due to an issue with uncompressed IP bitmap in core:
> // xref: https://github.com/kubernetes/kubernetes/issues/44918
> // The node subnet mask size must be no more than pod subnet mask size + 16

File: cmd/kubeadm/app/phases/controlplane/manifests.go (hunk 2 above)
Comment (neolit123) [reply cmd/kubeadm/app/phases/controlplane/manifests.go:368]: this passes the IPv6DualStack feature gate to the KCM by default, so i'd say it's needed.
it is passed to the API server and other components too:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/controlplane/manifests.go#L197

File: cmd/kubeadm/app/phases/controlplane/manifests.go (hunk 3 above)
Comment (neolit123) [reply cmd/kubeadm/app/phases/controlplane/manifests.go:370]: i guess this is true, but we need to double check that we are backwards compatible.

the docs for the flag `--node-cidr-mask-size` say:
> Mask size for node cidr in cluster. Default is 24 for IPv4 and 64 for IPv6.

https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/

so i guess if the user is not passing the flag explicitly it should be good for them and with kubeadm's default podSubNet (that has `/16`)

**Review by neolit123 — COMMENTED** [review]
thanks a lot @aojea i did a first pass and overall this looks fine.
but we need to make sure we are not breaking the existing users.

/cc @Arvinderpal 
can you have a look at this change please?

**Comment by neolit123** [comment]
> kubeadm validates the podSubnet against the corresponding cluster node-cidr-masks of the controller-manager, failing if they are not correct

i'd reword the release note to:
> kubeadm: improve the validation of podSubnet against the corresponding cluster "--node-cidr-mask-size" of the kube-controller-manager, fail if the values are not compatible

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go [hunk 8]
```diff
@@ -403,21 +403,45 @@ func ValidateIPNetFromString(subnetStr string, minAddrs int64, isDualStack bool,
 	return allErrs
 }
 
+// ValidatePodSubnetNodeMask validates that the relation between podSubnet and node-masks are correct
```
Comment (neolit123) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go:406]: `is correct`

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go (hunk 1 above)
Comment (neolit123) [reply cmd/kubeadm/app/apis/kubeadm/validation/validation.go:429]: it is not consistent with the rest of the functions though, was there an error if `c` was used for this function?

Comment (aojea) [reply cmd/kubeadm/app/apis/kubeadm/validation/validation.go:429]: Nope, will change it again, :)

File: cmd/kubeadm/app/phases/controlplane/manifests.go (hunk 2 above)
Comment (aojea) [reply cmd/kubeadm/app/phases/controlplane/manifests.go:368]: Ahh, I got it now, let me change it

File: cmd/kubeadm/app/phases/controlplane/manifests.go (hunk 3 above)
Comment (aojea) [reply cmd/kubeadm/app/phases/controlplane/manifests.go:370]: Yeah, for validation we use the user configured, or the default depending on the IP family and the dual stack flag.
We don't need to pass the default to the controller manager because is redundant

File: cmd/kubeadm/app/componentconfigs/controllermanager.go (hunk 4 above)
Comment (aojea) [reply cmd/kubeadm/app/componentconfigs/controllermanager.go:36]: Yeah, I was wondering what will be the best place, I choose the wrong one :)

File: cmd/kubeadm/app/phases/controlplane/manifests.go (hunk 3 above)
Comment (neolit123) [reply cmd/kubeadm/app/phases/controlplane/manifests.go:370]: though this seems like a breaking change to users that are using IPV6 and don't know how to calculate the IPV6 masks. 
the default value is 64 there and they will get validation error now.

```
case podNetSize == 112:
  // Special case, allows 256 nodes, 256 pods/node
  nodeCidrSize = 120
case podNetSize < 112:
  // Use multiple of 8 for node CIDR, with 512 to 64K nodes
 nodeCidrSize = totalBits - ((totalBits-podNetSize-1)/8-1)*8
```

so previously if they were passing `2001:db8:1::/112` for podsubnet they would get `--node-cidr-mask-size=120` and if they were passing `some-ipv6-ip/111`, they would get `--node-cidr-mask-size=128`.

given ipv6 support in k8s is shaky due to the bitmap problems we might go with this and add `action required:` in the release note...

**Comment by aojea** [comment]
> with the `release-note` annotation.

it has the annotation, but it always renders this way to me

**Comment by aojea** [comment]
checking the controller-manager logs this is passing 0 to the flags
```
2020-10-20T11:24:08.60783726Z stderr F I1020 11:24:08.607371       1 flags.go:59] FLAG: --node-cidr-mask-size="0"
2020-10-20T11:24:08.607842428Z stderr F I1020 11:24:08.607376       1 flags.go:59] FLAG: --node-cidr-mask-size-ipv4="0"
2020-10-20T11:24:08.607846718Z stderr F I1020 11:24:08.607381       1 flags.go:59] FLAG: --node-cidr-mask-size-ipv6="0"
```

it this expected @neolit123

**Comment by neolit123** [comment]
> checking the controller-manager logs this is passing 0 to the flags

`/cmd/kubeadm/app/phases/controlplane/manifests.go` is the only place where `--node-cidr-mask-size` is managed by kubeadm.
https://github.com/kubernetes/kubernetes/search?q=path%3A%2Fcmd%2Fkubeadm+%22node-cidr-mask-size%22

so i wonder if the KCM defaults to 0 if the flag is not passed - i.e. the KCM docs are wrong about it defaulting to 24 / 64?

**Comment by aojea** [comment]
/hold

this is breaking ipv6 and dual

**Comment by neolit123** [comment]
> this is breaking ipv6 and dual

did you find out why?

**Comment by aojea** [comment]
> if this causing problems, maybe the 0 value is then updated to 24 / 64 in the KCM code?

it is 

https://github.com/kubernetes/kubernetes/blob/f2e23afcf13fef805ad5341bd7a5625de5982d6c/cmd/kube-controller-manager/app/core.go#L631-L633

> did you find out why?

_looking_

**Comment by aojea** [comment]
> _looking_

found it https://github.com/kubernetes-sigs/kind/pull/1903

**Comment by aojea** [comment]
@neolit123 @Arvinderpal I've incorporated the service subnet validation to this pr so we merge both at the same time and we don't have to wait for other sigs to approve docs only changes. PTAL

File: cmd/kubeadm/app/phases/controlplane/manifests.go (hunk 3 above)
Comment (aojea) [reply cmd/kubeadm/app/phases/controlplane/manifests.go:370]: in the other hand, you see what happen with kind, we were changing a configuration option without telling the user.
The user has to known in advance how to subnet their cluster, the CNI will require this info.

**Comment by aojea** [comment]
/retest
unrelated failures
/test pull-kubernetes-e2e-kind-dual-canary

**Comment by aojea** [comment]
it's ok now, the dual-canary fails because is not using latest kind version

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go [hunk 9]
```diff
@@ -403,6 +403,81 @@ func ValidateIPNetFromString(subnetStr string, minAddrs int64, isDualStack bool,
...
+		mask := podSubnet.Mask
+		maskSize, _ := mask.Size()
+		// obtain node-cidr-mask
+		nodeMask := getClusterNodeMask(c, utilnet.IsIPv6(podSubnet.IP))
+		// the pod subnet mask needs to allow one or multiple node-masks
+		// i.e. if it has a /24 the node mask must be between 24 and 32 for ipv4
+		if maskSize > nodeMask {
+			allErrs = append(allErrs, field.Invalid(fldPath, podSubnet.String(), "pod subnet size is smaller than the node subnet size"))
+		} else if (nodeMask - maskSize) > constants.PodSubnetNodeMaskMaxDiff {
+			allErrs = append(allErrs, field.Invalid(fldPath, podSubnet.String(), "pod subnet mask and node-mask difference can not be greater than 16"))
```
Comment (Arvinderpal) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go:441]: nit: replace 16 with PodSubnetNodeMaskMaxDiff

File: cmd/kubeadm/app/phases/controlplane/manifests.go (hunk 3 above)
Comment (Arvinderpal) [reply cmd/kubeadm/app/phases/controlplane/manifests.go:370]: The previous experience of kubeadm making the change for the user had it's problems. It generally lead to confusion, so I'm +1 for making this change. 
However, it will be breaking change for some users. I'm okay with a release note documenting this.

**Review by Arvinderpal — COMMENTED** [review]
Just a couple of comments. Overall, looks good.

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go (hunk 9 above)
Comment (aojea) [reply cmd/kubeadm/app/apis/kubeadm/validation/validation.go:441]: yep

**Comment by neolit123** [comment]
> action required: kubeadm no longer sets the node-mask automatically on IPv6 deployments, you must check that your IPv6 service subnet mask is compatible with the default node mask /64 or set it accordenly

about this part of the release note. i'm wondering if we should give more details about the <= 112 handling from before. what do you think?

**Comment by aojea** [comment]
> about this part of the release note. i'm wondering if we should give more details about the <= 112 handling from before. what do you think?

done, but not easy to explain though

**Comment by neolit123** [comment]
> done, but not easy to explain though

it also modified 112->120, should we mention that too, or just tell the users to be explicit?

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go [hunk 10]
```diff
@@ -403,6 +403,78 @@ func ValidateIPNetFromString(subnetStr string, minAddrs int64, isDualStack bool,
...
+		if maskSize > nodeMask {
+			allErrs = append(allErrs, field.Invalid(fldPath, podSubnet.String(), "pod subnet size is smaller than the node subnet size"))
+		} else if (nodeMask - maskSize) > constants.PodSubnetNodeMaskMaxDiff {
+			allErrs = append(allErrs, field.Invalid(fldPath, podSubnet.String(), fmt.Sprintf("pod subnet mask and node-mask difference can not be greater than %d", constants.PodSubnetNodeMaskMaxDiff)))
+		}
+	}
+	return allErrs
+}
+
+// getClusterNodeMask return the corresponding node-cidr-mask
```
Comment (neolit123) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go:447]: return -> returns

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go [hunk 11]
```diff
@@ -403,6 +403,78 @@ func ValidateIPNetFromString(subnetStr string, minAddrs int64, isDualStack bool,
...
+		maskArg = "node-cidr-mask-size-ipv6"
+	} else if isDualStack && !isIPv6 {
+		maskArg = "node-cidr-mask-size-ipv4"
+	} else {
+		maskArg = "node-cidr-mask-size"
+	}
+
+	if v, ok := c.ControllerManager.ExtraArgs[maskArg]; ok && v != "" {
+		// assume it is an integer, if not it will fail later
+		maskSize, _ = strconv.Atoi(v)
```
Comment (neolit123) [inline cmd/kubeadm/app/apis/kubeadm/validation/validation.go:469]: i guess it may fail later in `ValidatePodSubnetNodeMask` for this check `} else if (nodeMask - maskSize) > constants.PodSubnetNodeMaskMaxDiff {` or can this pass and then later fail on the KCM level?

i think it's worth making `getClusterNodeMask` return `(int, error)` and returning the error from Atoi:
```
if err != nil {
    errors.Wrapf(err, "could not parse the value of the kube-controller-manager flag %s as an integer", maskArg)
}
```

**Comment by aojea** [comment]
> it also modified 112->120, should we mention that too, or just tell the users to be explicit?

 users with /112 will fail in validation and are warned by `your IPv6 service subnet mask is compatible with the default node mask /64`.

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go (hunk 11 above)
Comment (neolit123) [reply cmd/kubeadm/app/apis/kubeadm/validation/validation.go:469]: i guess there could be a unit test for that in TestGetClusterNodeMask too.

File: cmd/kubeadm/app/constants/constants.go [hunk 12]
```diff
@@ -199,6 +199,20 @@ const (
 	// We need at least ten, because the DNS service is always at the tenth cluster clusterIP
 	MinimumAddressesInServiceSubnet = 10
 
+	// MaximumBitsForServiceSubnet defines maximum possible size of the service subnet in terms of bits.
+	// For example, if value is 20, then the largest supported service subnet is /12 for IPv4 and /108 for IPv6.
```
Comment (neolit123) [inline cmd/kubeadm/app/constants/constants.go:203]: if value -> if the value

File: cmd/kubeadm/app/constants/constants.go [hunk 13]
```diff
@@ -199,6 +199,20 @@ const (
...
+	// Note however that anything in between /108 and /112 will be clamped to /112 due to the limitations of the underlying allocation logic.
```
Comment (neolit123) [inline cmd/kubeadm/app/constants/constants.go:204]: did we had a ticket for that in k/k that we can link to from here?
maybe have a `TODO: ...` to update this once/if the bitmap problems are resolved?

**Comment by neolit123** [comment]
added final comments, i think this is good!

File: cmd/kubeadm/app/constants/constants.go (hunk 13 above)
Comment (aojea) [reply cmd/kubeadm/app/constants/constants.go:204]: added the kep that explains the whole thing

File: cmd/kubeadm/app/constants/constants.go (hunk 12 above)
Comment (aojea) [reply cmd/kubeadm/app/constants/constants.go]: done

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go (hunk 11 above)
Comment (aojea) [reply cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: done

File: cmd/kubeadm/app/apis/kubeadm/validation/validation.go (hunk 10 above)
Comment (aojea) [reply cmd/kubeadm/app/apis/kubeadm/validation/validation.go]: done

**Comment by neolit123** [comment]
please squash the last commit `constant comments` into `kubeadm validate maximum service subnet size`

/approve

**Comment by neolit123** [comment]
thanks @aojea !
the change seems fine to me and with @Arvinderpal giving his +1, i think we should merge this for 1.20.

/lgtm
/retest
(bazel debian pull flake?)

**Comment by aojea** [comment]
/test pull-kubernetes-bazel-test

//pkg/kubelet:go_default_test                                            FAILED in 1 out of 2 in 12.3s

File: cmd/kubeadm/app/phases/controlplane/manifests.go (hunk 3 above)
Comment (aoxn) [reply cmd/kubeadm/app/phases/controlplane/manifests.go:370]: > 

@Arvinderpal How does user recover from this breaking change then?

Comment (aojea) [reply cmd/kubeadm/app/phases/controlplane/manifests.go:370]: what do you mean exactly @aoxn ?

the change was done in an alpha feature in kubernetes 1.20 so the change is justified, are you upgrading from 1.19?

 you need to set the `--node-cidr-mask-size=` accordingly if the default value doesn't work for you, which part is failing to you?
